File: packages/gh-tasks/src/lib/scope.ts
import { execFileSync } from 'node:child_process';

export type Scope = 'repo' | 'org' | 'user';

const VALID: readonly Scope[] = ['repo', 'org', 'user'];

export class ScopeError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'ScopeError';
  }
}

export interface DetectScopeOptions {
  argv: readonly string[];
  hasGitRemote?: () => boolean;
  config?: { defaultScope?: Scope };
}

/**
 * Resolve the working scope.
 *
 * Order:
 *   1. `--scope repo|org|user` flag
 *   2. git remote `origin` exists → `repo`
 *   3. `~/.config/ozzylabs/gh-tasks.toml` `default_scope` (passed via `config`)
 *   4. fallback → `user`
 */
export function detectScope(opts: DetectScopeOptions): Scope {
  const fromFlag = parseScopeFlag(opts.argv);
  if (fromFlag) return fromFlag;

  const detector = opts.hasGitRemote ?? defaultHasGitRemote;
  if (detector()) return 'repo';

  if (opts.config?.defaultScope) return opts.config.defaultScope;

  return 'user';
}

export function parseScopeFlag(argv: readonly string[]): Scope | null {
  for (let i = 0; i < argv.length; i++) {
    const arg = argv[i];
    if (arg === undefined) continue;

    if (arg.startsWith('--scope=')) {
      const value = arg.slice('--scope='.length);
      return assertScope(value);
    }
    if (arg === '--scope') {
      const next = argv[i + 1];
      if (next === undefined) {
        throw new ScopeError('--scope フラグに値が指定されていません');
      }
      return assertScope(next);
    }
  }
  return null;
}

function assertScope(value: string): Scope {
  if ((VALID as readonly string[]).includes(value)) {
    return value as Scope;
  }
  throw new ScopeError(`不正な --scope 値: '${value}' (有効値: ${VALID.join(' | ')})`);
}

function defaultHasGitRemote(): boolean {
  try {
    execFileSync('git', ['remote', 'get-url', 'origin'], { stdio: 'pipe' });
    return true;
  } catch {
    return false;
  }
}

File: internal/scope/scope.go
// Package scope provides the 3-scope abstraction (repo / org / user).
package scope

import (
	"github.com/ozzy-labs/gh-tasks/internal/i18n"
)

// Scope is the working scope.
type Scope string

// Recognized scope values.
const (
	Repo Scope = "repo"
	Org  Scope = "org"
	User Scope = "user"
)

// Valid lists the accepted scopes in canonical order.
var Valid = []Scope{Repo, Org, User}

// ScopeError is returned when a --scope flag is missing a value or has an
// unrecognized value.
//
// Use errors.As(err, &target) to test for this type:
//
//	var se *scope.ScopeError
//	if errors.As(err, &se) { ... }
type ScopeError struct{ i18n.Payload }

// Error satisfies the error interface.
func (e *ScopeError) Error() string { return e.Key }

func newError(key string, args ...any) *ScopeError {
	return &ScopeError{Payload: i18n.NewPayload(key, args...)}
}

// DetectOptions configures Detect.
//
// Flag carries the raw string value of the --scope flag as parsed by cobra
// (empty string means the flag was not supplied). The previous Argv-based
// scanner was retired in favour of cobra-authoritative flag handling.
type DetectOptions struct {
	Flag         string
	HasGitRemote func() bool
	DefaultScope Scope
}

// Detect resolves the working scope.
//
// Order:
//  1. Flag (cobra-parsed --scope value)
//  2. git remote origin exists → repo
//  3. config DefaultScope
//  4. fallback → user
func Detect(opts DetectOptions) (Scope, error) {
	if opts.Flag != "" {
		return assertScope(opts.Flag)
	}
	if opts.HasGitRemote != nil && opts.HasGitRemote() {
		return Repo, nil
	}
	if opts.DefaultScope != "" {
		return opts.DefaultScope, nil
	}
	return User, nil
}

func assertScope(v string) (Scope, error) {
	for _, s := range Valid {
		if string(s) == v {
			return s, nil
		}
	}
	return "", newError("error.scope.invalid", "value", v, "valid", i18n.JoinPipe(Valid))
}
